ghc-lib-gen/src/Main.hs

```diff
@@ -65,11 +66,17 @@ ghcLibParserIncludeDirs =
 -- adjusting.
 ghcLibParserHsSrcDirs :: [Cabal] -> [FilePath]
 ghcLibParserHsSrcDirs lib =
-  nubSort $
-   [ "ghc-lib/stage0/compiler/build"
-   , "ghc-lib/stage1/compiler/build"] ++
-   map takeDirectory cabalFileLibraries ++
-   askFiles lib "hs-source-dirs:"
+  -- Sort by length so the longest paths are at the front. We do this
+  -- so that in 'calcParserModules', longer substituions are performed
+  -- before shorter ones (and bad things will happen if that were
+  -- not the case).
+  sortBy (flip (comparing length))
+    ([ "ghc-lib/stage0/compiler/build"
+     , "ghc-lib/stage1/compiler/build"
+     , "ghc-lib/stage0/libraries/ghci/build"
+     , "ghc-lib/stage0/libraries/ghc-heap/build"
+     ] ++ map takeDirectory cabalFileLibraries ++
+      askFiles lib "hs-source-dirs:")
```

Comment:
I'd definitely `$` this whole block rather than having surrounding brackets that mess up indentation etc.